userd/autostart.go

```diff
@@ -63,30 +71,81 @@ func expandDesktopFields(in string) string {
 	return string(out)
 }
 
-// Note: consider wrappers/desktop.go if we start parsing more than Exec line
-func findExec(desktopFileContent []byte) (string, error) {
-	scanner := bufio.NewScanner(bytes.NewBuffer(desktopFileContent))
-	execCmd := ""
+type skipDesktopFileError struct {
+	reason string
+}
+
+func (s *skipDesktopFileError) Error() string {
+	return s.reason
+}
+
+func isOneOfIn(of []string, other []string) bool {
+	for _, one := range of {
+		if strutil.ListContains(other, one) {
+			return true
+		}
+	}
+	return false
+}
+
+func loadAutostartDesktopFile(path string) (command string, err error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
 	for scanner.Scan() {
 		bline := scanner.Bytes()
-
-		if !bytes.HasPrefix(bline, []byte("Exec=")) {
+		if bytes.HasPrefix(bline, []byte("#")) {
 			continue
 		}
-
-		full := string(bline[len("Exec="):])
-		execCmd = expandDesktopFields(full)
-		break
+		split := bytes.SplitN(bline, []byte("="), 2)
+		if len(split) != 2 {
+			continue
+		}
+		// See https://standards.freedesktop.org/autostart-spec/autostart-spec-latest.html
+		// for details on how Hidden, OnlyShowIn, NotShownIn are handled.
+		switch string(split[0]) {
+		case "Exec":
+			command = strings.TrimSpace(expandDesktopFields(string(split[1])))
+		case "Hidden":
+			if bytes.Equal(split[1], []byte("true")) {
+				return "", &skipDesktopFileError{"desktop file is hidden"}
+			}
+		case "OnlyShowIn":
+			onlyIn := splitSkippingEmpty(string(split[1]), ';')
+			if !isOneOfIn(currentDesktop, onlyIn) {
+				return "", &skipDesktopFileError{fmt.Sprintf("current desktop %q not included in %q", currentDesktop, onlyIn)}
+			}
+		case "NotShownIn":
+			notIn := splitSkippingEmpty(string(split[1]), ';')
+			if isOneOfIn(currentDesktop, notIn) {
+				return "", &skipDesktopFileError{fmt.Sprintf("current desktop %q excluded by %q", currentDesktop, notIn)}
+			}
+		case "X-GNOME-Autostart-enabled":
```

Comment:
I "love" how all but the last word in the key name are capitalised. Consistency FTW